src/rules/selector-max-depth/index.js

```diff
@@ -0,0 +1,78 @@
+import resolvedNestedSelector from "postcss-resolve-nested-selector"
+
+import {
+  isStandardRule,
+  isStandardSelector,
+  report,
+  ruleMessages,
+  validateOptions,
+} from "../../utils"
+
+export const ruleName = "selector-max-depth"
+
+export const messages = ruleMessages(ruleName, {
+  expected: (selector, depth) => `Expected "${selector}" to have a depth equal to or less than ${depth}`,
+})
+
+export default function (max) {
+  return (root, result) => {
+    const innerSelectorReg = /(?::not\((.*?)\))/g
+    const validOptions = validateOptions(result, ruleName, {
+      actual: max,
+      possible: [function (max) {
+        return typeof max === "number" && max > 0
+      }],
+    })
+    if (!validOptions) { return }
+
+    // Compares depth of a selector with the maximum
+    function checkDepth(selector, rule) {
+      // a) Replacing all :not() occureces with a dummy selector
+      const stripped = selector.replace(innerSelectorReg, "a")
+        // b) Replacing "[...]", "(...)", "+", "~" (the last two might come with
```

Comment:
I'm afraid I have little to add to the post above for justification. At first the intention was to implement the limit to the number of HTML levels in a selector, much like nesting, but more strictly. In fact, I am not willing to defend this definition (`+` and `~` are different than ` ` and `>`) till the very end, as I think the more logacal it would be the better fore everyone. Moreover, now that I think of it, `.a + .b + .c + .d + .e + .f` (if there is a preson savage enough to come up with such a selector :) ) is no less destructive.

Just one thing, it's much more of an issue of coupling a selector with HTML than of specificity.